### PR TITLE
Update DCRA for private linked cluster and create new DCE when cluster and workspace do not match

### PR DESCRIPTION
**Pull request by Sohamdg081992**
[comment]: # (Note that your PR title should follow the conventional commit format: https://conventionalcommits.org/en/v1.0.0/#summary)
# PR Description
This change updates DCRA for private linked cluster based on flag and create new DCE when cluster and workspace do not match.
The manual linking for the DCE to the cluster resource is no longer needed after this change.
After manual linking , a PUT request is made to create a DCRA with name "configurationAccessEndpoint".
This change makes sure this new DCRA(configurationAccessEndpoint) is created and linking is done only if useAzureMonitorPrivateLinkScope flag is enabled and if DCE and cluster are in same region.
Below are screenshots of the linking:

Step showing network capture while manually adding DCRA
priavatelinktest

Step showing creation of DCRA . Name defaults to configurationAccessEndpoint -> [doc](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/monitor_data_collection_rule_association).
priavatelinktest1

Step showing creation of DCRA in portal.
priavatelinktest2

[comment]: # (The below checklist is for PRs adding new features. If a box is not checked, add a reason why it's not needed.)
# New Feature Checklist

- [ ] List telemetry added about the feature.
- [ ] Link to the one-pager about the feature.
- [ ] List any tasks necessary for release (3P docs, AKS RP chart changes, etc.) after merging the PR.
- [ ] Attach results of scale and perf testing.

[comment]: # (The below checklist is for code changes. Not all boxes necessarily need to be checked. Build, doc, and template changes do not need to fill out the checklist.)
# Tests Checklist

- [ ] Have end-to-end Ginkgo tests been run on your cluster and passed? To bootstrap your cluster to run the tests, follow [these instructions](/otelcollector/test/README.md#bootstrap-a-dev-cluster-to-run-ginkgo-tests).
  - Labels used when running the tests on your cluster:
    - [ ] `operator`
    - [ ] `windows`
    - [ ] `arm64`
    - [ ] `arc-extension`
    - [ ] `fips`
- [ ] Have new tests been added? For features, have tests been added for this feature? For fixes, is there a test that could have caught this issue and could validate that the fix works?
  - [ ] Is a new scrape job needed?
    - [ ] The scrape job was added to the folder [test-cluster-yamls](/otelcollector/test/test-cluster-yamls/) in the correct configmap or as a CR. 
  - [ ] Was a new test label added?
    - [ ] A string constant for the label was added to [constants.go](/otelcollector/test/utils/constants.go).
    - [ ] The label and description was added to the [test README](/otelcollector/test/README.md).
    - [ ] The label was added to this [PR checklist](/.github/pull_request_template).
    - [ ] The label was added as needed to [testkube-test-crs.yaml](/otelcollector/test/testkube/testkube-test-crs.yaml).
  - [ ] Are additional API server permissions needed for the new tests?
    - [ ] These permissions have been added to [api-server-permissions.yaml](/otelcollector/test/testkube/api-server-permissions.yaml).
  - [ ] Was a new test suite (a new folder under `/tests`) added?
    - [ ] The new test suite is included in [testkube-test-crs.yaml](/otelcollector/test/testkube/testkube-test-crs.yaml).

### CODE REVIEW

File: AddonTerraformTemplate/main.tf [hunk 1]
```diff
@@ -90,6 +90,18 @@ resource "azurerm_monitor_data_collection_rule_association" "dcra" {
   ]
 }
 
+resource "azurerm_monitor_data_collection_rule_association" "dcraprivatelink" {
+  count = var.useAzureMonitorPrivateLinkScope && azurerm_kubernetes_cluster.k8s.location == azurerm_monitor_workspace.amw.location ? 1 : 0
+  data_collection_endpoint_id = azurerm_monitor_data_collection_endpoint.dce.id
+  target_resource_id  = azurerm_kubernetes_cluster.k8s.id
+  description             = "Association of data collection rule. Deleting this association will break the data collection for this AKS Cluster."
```
Comment (Sohamdg081992) [inline AddonTerraformTemplate/main.tf:97]: Name defaults to configurationAccessEndpoint -> [doc](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/monitor_data_collection_rule_association).

File: AddonTerraformTemplate/main.tf [hunk 2]
```diff
@@ -90,6 +90,18 @@ resource "azurerm_monitor_data_collection_rule_association" "dcra" {
   ]
 }
 
+resource "azurerm_monitor_data_collection_rule_association" "dcraprivatelink" {
+  count = var.useAzureMonitorPrivateLinkScope && azurerm_kubernetes_cluster.k8s.location == azurerm_monitor_workspace.amw.location ? 1 : 0
```
Comment (Sohamdg081992) [inline AddonTerraformTemplate/main.tf:94]: To achieve this, you need to add a conditional check in your Terraform code to verify if the `azurerm_kubernetes_cluster` is in the same region as the `azurerm_monitor_workspace`. If the condition is met, update the `azurerm_monitor_data_collection_rule_association` name accordingly.

Here is the updated code:

```terraform
variable "useAzureMonitorPrivateLinkScope" {
  description = "Flag to determine if Azure Monitor Private Link Scope should be used"
  type        = bool
  default     = false
}

resource "azurerm_monitor_data_collection_rule_association" "example" {
  count = var.useAzureMonitorPrivateLinkScope && azurerm_kubernetes_cluster.example.location == azurerm_monitor_workspace.example.location ? 1 : 0

  name                = "${azurerm_kubernetes_cluster.example.id}/providers/microsoft.insights/dataCollectionRuleAssociations/configurationAccessEndpoint"
  data_collection_rule_id = azurerm_monitor_data_collection_rule.example.id
  target_resource_id  = azurerm_kubernetes_cluster.example.id
}
```

### Explanation:
1. **Variable Definition**: A variable `useAzureMonitorPrivateLinkScope` is defined to control the conditional logic.
2. **Resource Count**: The `count` parameter is used to conditionally create the `azurerm_monitor_data_collection_rule_association` resource. It checks if `useAzureMonitorPrivateLinkScope` is true and if the `azurerm_kubernetes_cluster` is in the same region as the `azurerm_monitor_workspace`.
3. **Resource Name**: The `name` parameter is updated to use the `azurerm_kubernetes_cluster` resource ID concatenated with `"providers/microsoft.insights/dataCollectionRuleAssociations/configurationAccessEndpoint"`.

This ensures that the `azurerm_monitor_data_collection_rule_association` resource is only created if the conditions are met, and the name is updated accordingly.

Comment (bragi92) [inline AddonTerraformTemplate/main.tf:94]: What about the scenario where the region does not match? Do you want to request the customer for a 
DCE id as a parameter maybe and populate this field accordingly?

File: AddonTerraformTemplate/variables.tf [hunk 3]
```diff
@@ -37,3 +37,8 @@ variable "resource_group_location" {
   default     = "eastus"
   description = "Location of the resource group."
 }
+
+variable "useAzureMonitorPrivateLinkScope" {
```
Comment (bragi92) [inline AddonTerraformTemplate/variables.tf:41]: Why is this only being done for terraform? Can we include bicep, arm, policy too in this?

Comment (Sohamdg081992) [reply AddonTerraformTemplate/variables.tf:41]: Thanks Kaveesh! I am doing this change for terraform currently to mitigate the issue. I will do the other template changes after hearing back from PMs on the scenarios.

Comment (bragi92) [reply AddonTerraformTemplate/variables.tf:41]: I'm not sure I fully understand, what is the question that you're waiting to hear back from the PMs? Don't the other onboarding scenarios also need to be mitigated/have the same scenario which is onboarding a private AKS cluster?

We should keep the templates consistent across onboarding mechanisms.

File: AddonTerraformTemplate/main.tf (hunk 2 above)
Comment (Sohamdg081992) [reply AddonTerraformTemplate/main.tf:94]: Thanks Kaveesh! Yes that makes sense. let me update the template.

Comment (Sohamdg081992) [reply AddonTerraformTemplate/main.tf]: Updated in the latest commit.

**Review by bragi92 — APPROVED** [review]

**Pull request closed without merging**